snakemake/dag.py

```diff
@@ -937,6 +937,7 @@ def update_needrun(job):
 
         queue = list(filter(reason, candidates))
         visited = set(queue)
+        candidates_set = set(candidates)
```

Comment:
Good catch, checking existence in a list was of course not intentional. Is there any reason not to use a set directly in line 924?

Reply:
The code before this requires the jobs to be in BFS order, so it still requires the list.